src/features/defi/providers/cosmos/components/CosmosManager/Withdraw/CosmosWithdraw.tsx

```diff
@@ -0,0 +1,148 @@
+import { useWallet } from 'hooks/useWallet/useWallet'
+import {
+  MergedActiveStakingOpportunity,
+  useCosmosSdkStakingBalances,
+} from 'pages/Defi/hooks/useCosmosSdkStakingBalances'
+import { selectAssetById, selectMarketDataById } from 'state/slices/selectors'
+import { useAppSelector } from 'state/store'
+
+import { Confirm } from './components/Confirm'
+import { Status } from './components/Status'
+import { Withdraw } from './components/Withdraw'
+import { CosmosWithdrawActionType } from './WithdrawCommon'
+import { WithdrawContext } from './WithdrawContext'
+import { initialState, reducer } from './WithdrawReducer'
+
+export const CosmosWithdraw = () => {
+  const translate = useTranslate()
+  const [state, dispatch] = useReducer(reducer, initialState)
+  const { query, history, location } = useBrowserRouter<DefiQueryParams, DefiParams>()
+  const { chainId, contractAddress, assetReference } = query
+
+  const assetNamespace = 'slip44' // TODO: add to query, why do we hardcode this?
+  // Asset info
+  const assetId = toAssetId({
+    chainId,
+    assetNamespace,
+    assetReference, // TODO: handle multiple denoms
+  })
+  const underlyingAssetId = toAssetId({
+    // TODO: Underlying asset is the same as the staked asset for now, handle multiple denoms
+    chainId,
+    assetNamespace,
+    assetReference,
+  })
+  const asset = useAppSelector(state => selectAssetById(state, assetId))
+  const underlyingAsset = useAppSelector(state => selectAssetById(state, underlyingAssetId))
+  const marketData = useAppSelector(state => selectMarketDataById(state, assetId))
+  const feeAssetId = toAssetId({
+    chainId,
+    assetNamespace,
+    assetReference,
+  })
+  const feeMarketData = useAppSelector(state => selectMarketDataById(state, feeAssetId))
+
+  // user info
+  const chainAdapterManager = getChainAdapterManager()
+  const chainAdapter = chainAdapterManager.get(chainId)
+  const { state: walletState } = useWallet()
+
+  const opportunities = useCosmosSdkStakingBalances({ assetId })
+  const cosmosOpportunity = useMemo(
+    () =>
+      opportunities?.cosmosSdkStakingOpportunities?.find(
+        opportunity => opportunity.address === contractAddress,
+      ) ?? {},
+    [opportunities, contractAddress],
+  ) as unknown as MergedActiveStakingOpportunity // TODO: remove casting
+
+  useEffect(() => {
+    ;(async () => {
+      try {
+        if (!(walletState.wallet && contractAddress && chainAdapter)) return
+        const address = await chainAdapter.getAddress({ wallet: walletState.wallet })
+
+        dispatch({
+          type: CosmosWithdrawActionType.SET_USER_ADDRESS,
+          payload: address,
+        })
+        dispatch({
+          type: CosmosWithdrawActionType.SET_OPPORTUNITY,
+          payload: { ...cosmosOpportunity },
+        })
+      } catch (error) {
+        // TODO: handle client side errors
+        console.error('CosmosWithdraw error:', error)
```

Comment:
moduleLogger

Reply:
Also from the Foxy implementation, will add it in the same stacked PR

Reply:
https://github.com/shapeshift/web/pull/2323